app-libs/exchange-oracle/src/mock.rs

```diff
@@ -0,0 +1,99 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::Error, exchange_rate_oracle::OracleSource, metrics_exporter::ExportMetrics, TradingPair,
+};
+use itc_rest_client::{http_client::HttpClient, rest_client::RestClient};
+use itp_types::ExchangeRate;
+use std::{
+	time::{Duration, Instant},
+	vec::Vec,
+};
+use url::Url;
+
+/// Mock metrics exporter.
+#[derive(Default)]
+pub(crate) struct MetricsExporterMock {
+	number_requests: RwLock<u64>,
+	response_times: RwLock<Vec<u128>>,
+	exchange_rates: RwLock<Vec<(TradingPair, ExchangeRate)>>,
+}
+
+impl MetricsExporterMock {
+	pub fn get_number_request(&self) -> u64 {
+		*self.number_requests.read().unwrap()
+	}
+
+	pub fn get_response_times(&self) -> Vec<u128> {
+		self.response_times.read().unwrap().clone()
+	}
+
+	pub fn get_exchange_rates(&self) -> Vec<(TradingPair, ExchangeRate)> {
+		self.exchange_rates.read().unwrap().clone()
+	}
+}
+
+impl ExportMetrics for MetricsExporterMock {
+	fn increment_number_requests(&self, _source: String) {
+		(*self.number_requests.write().unwrap()) += 1;
+	}
+
+	fn record_response_time(&self, _source: String, timer: Instant) {
+		self.response_times.write().unwrap().push(timer.elapsed().as_millis());
+	}
+
+	fn update_exchange_rate(
+		&self,
+		_source: String,
+		exchange_rate: ExchangeRate,
+		trading_pair: TradingPair,
+	) {
+		self.exchange_rates.write().unwrap().push((trading_pair, exchange_rate));
+	}
+}
+
+/// Mock oracle source.
+pub(crate) struct OracleSourceMock;
```

Comment:
Same for the above - though that might be a spy already?

Reply:
In Rust we don't ever have Mocks in the CleanCoder definition, all of them are stubs. But since substrate also calls them mocks, I think we're kind of consistent. Otherwise we'd have to re-name all our mocks to stubs.. ?

Reply:
As for the spy: We fully stub/mock the implementation here, and if I understand correctly, a spy would be a partial mocking.